### ISSUE TITLE
Proposal: Support for Workflow Run status update

### ISSUE THREAD

**Issue by susheel** [ga4gh/workflow-execution-service-schemas#133]
As part of the asynchronous [`callback_url` proposal for TES](https://github.com/ga4gh/task-execution-schemas/issues/121), we would like to have a mechanism to allow TES endpoints to POST status updates back to a WES endpoint.

Requirements:

- Need for a `POST /runs/{run_id}/status` as a mechanism that will accept a TesResponse
- TES will use the submission auth headers as part of the POST request

**Comment by jaeddy** [comment]
I think it might make sense for this to be a `PUT` operation. I believe most WES implementations would create some initial `status` whenever a new `run` is created (and before sending anything to TES). The callback would then be used to _update_ the status.

**Comment by susheel** [comment]
Happy with `PUT` or `POST`. I suggested `POST` as it was similar to webhook conventions

**Comment by uniqueg** [comment]
Since not every WES will be implementing a TES backend, it might perhaps be better to just add this as a recommended implementation detail for those implementers who wish to support this behavior. A reference to the relevant section in the TES specs might perhaps suffice, especially since any TES client will likely refer to the TES specs anyways.

**Comment by jaeddy** [comment]
@uniqueg thanks - now that various issues with the docs build system, bumping to OpenAPI 3, and integrating the new service-info API are getting close to finished, I'm hoping to look into the status callbacks as the next real substantive update to the spec. I appreciate the detailed proposals that @susheel and you have put together, and your additional notes in [this thread](https://github.com/ga4gh/cloud-interop-testing/issues/98#issuecomment-645485554). I'll take a closer look soon!

**Comment by uniqueg** [comment]
Thanks a lot @jaeddy, sounds great. Looks like OpenAPI3 supports the [definition of callbacks](https://swagger.io/docs/specification/callbacks/ ): So with the bump to v3, that's probably something that we could make use of. Please let me know if you need some help spec'cing that out, during the upcoming hackathon or otherwise.

**Comment by uniqueg** [comment]
Agreed. And this is also how we implemented the PoC showing how callbacks can be used.

**Comment by golharam** [comment]
I came across this issue as I also have a need for a PUT endpoint to update run status information.  I'm not using TES however, I'm using a different engine to execute workflows.  When the workflow completes, I don't have a way to update the run state in WES.  My current approach is to have a separate monitoring service that listens to workflow state changes from the engine.  Upon a state change,  the monitoring service will call an endpoint on the WES service to update the workflow state.  Right now I'm defining a separate API (/internal/callbacks/workflow-state-change), but really, I think this should be a PUT on /runs/{run_id}/status.

**Comment by patmagee** [comment]
@golharam I think WES needs to be looked at as an interoperability layer that sits in front of an execution environment instead of a standalone service that sits adjacent to the environment as a standalone service. The specifics of how WES is implemented are intentionally vague, since it makes no assumptions of whether it is just a translation layer on top of some other API, an engine itself, an independent service etc. To that end, the goal is always to keep the API footprint relatively small to cover these diverse needss

A workflow state change endpoint implies the WES itself is tracking state independently of whatever system it interfaces with and that state needs updating. This is more or less an implementation detail IMO and I think your best bet is to make a bespoke endpoint (which it seems like you to) that is off spec to meet your use case.
